content/articles/building-an-ensemble-learning-based-regression-model-using-python/index.md

```diff
@@ -0,0 +1,203 @@
+
+To select the useful features, run the following block of code:
+
+```python
+useful_features = [i for i in data.columns if i not in ("id", "target","kfold")]
+object_cols = [col for col in useful_features if "cat" in col]
+test_data = test_data[useful_features]
+```
+
+#### Modeling
+To build our model, we will run the following block of code:
+
+```python
+final_predictions =[]
+
+for fold in range(5):
+    xtrain = data[data.kfold != fold].reset_index(drop=True)
+    xvalid = data[data.kfold == fold].reset_index(drop=True)
+    xtest = test_data.copy()
+    
+    ytrain = xtrain.target
+    yvalid = xvalid.target
+    
+    xtrain = xtrain[useful_features]
+    xvalid = xvalid[useful_features]
+    
+    # Data Encoding 
+    oe = OrdinalEncoder()
+    xtrain[object_cols] = oe.fit_transform(xtrain[object_cols])
+    xvalid[object_cols] = oe.transform(xvalid[object_cols])
+    xtest[object_cols] = oe.transform(xtest[object_cols])
+    
+    # Model Training
+    model = XGBRegressor(random_state = fold, n_jobs=5)
+    model.fit(xtrain, ytrain)
+    preds_valid = model.predict(xvalid)
+    preds_test = model.predict(xtest)
+    final_predictions.append(preds_test)
+    print(fold, mean_squared_error(yvalid, preds_valid, squared=False))
+```
+
+For each fold, we will encode the data and then train the model using XGBoost (Extreme Gradient Boosting), an ensemble learning technique to boost the performance of our model.
+XGBoost is a regularized boosting technique that provides high predictive power and is faster than other boosting techniques. We will then evaluate each fold individually and print out the results of the model.
+
+#### Model evaluation
+After individually evaluating each fold, we will now evaluate our model's performance by getting the mean predictions on our test data. To do this, use the following code block:
+
+```python
+preds = np.mean(np.column_stack(final_predictions), axis=1)
+```
+
+To see how our model performed, we will output the results of our model's prediction using the following code:
+
+```python
+submission.target =preds
+submission.to_csv("submission1.csv", index=False)
+```
+
+>**Bonus:** You can submit a late submission to 30 days ML Kaggle challenge and see how your model performs, i.e., if you had signed up for the challenge earlier.
+
+### Hyperparameter optimization
+In this process, we'll fine-tune and optimize our model's algorithm parameters until we achieve the desired result.
+A few common XGBoost parameters with a large effect on the model perfomance include; **n_jobs**, **max_depth**, **learning_rate**, **n_estimators**, **colsample_bytree**, and **subsample**.
+
+To fine-tune our model, add the following changes to the XGBoost regressor:
+
+```python
+model = XGBRegressor(random_state = fold, n_jobs=5, learning_rate =0.1, subsample=0.8,
+                         max_depth = 5, min_child_weight = 1, gamma = 0, scale_pos_weight = 1)
+```
```

Comment:
What are the results from fine-tuning? Did the results improve? You've left us in the dark.

Reply:
@Bayler  I'll be adding a little content here, the one I've submitted already is not yet satisfactory for me. Besides that I think the rest are good

Reply:
@Bayler  I'll be adding a little content here, the one I've submitted already is not yet satisfactory for me. Besides that, I think the rest are good